Reagent dispensers

Moneyl opened the pull request:
Adds `ReagentDispenserComponent`. This expands upon #280 by adding a way to make machines which add/remove reagents from a solution (such as the chemical or booze dispensers). `ReagentDispenserComponent` was written in a way that it can be reused on multiple types of dispensers. This PR includes several entities using the component: `Chem dispenser`, `Booze dispenser`, and `Soda dispenser` and adds several new reagents to be dispensed from them. 

The reagent inventory of a dispenser is determined by defining and specifying a reagent pack in it's yaml prototype. See [chem_dispenser.yml](https://github.com/space-wizards/space-station-14/blob/63f5a08b60dc624943e5b0c02797d8ae32aa94ce/Resources/Prototypes/Entities/buildings/chem_dispenser.yml#L31) for an example of this.

This does not implement reactions, ingestion, or container color varying by solution.

Examples of the ui:
![Z77ohL1A3F](https://user-images.githubusercontent.com/8206401/65391741-bff72e80-dd3a-11e9-82c8-9b0203626b6b.gif)
![Robust Client_XYGw6cNaKV](https://user-images.githubusercontent.com/8206401/65391748-cb4a5a00-dd3a-11e9-899b-5e80976c0fc3.png)

Acruid commented on Content.Shared/Chemistry/Solution.cs:
```diff
@@ -17,6 +17,7 @@ public class Solution : IExposeData, IEnumerable<Solution.ReagentQuantity>
         // Most objects on the station hold only 1 or 2 reagents
         [ViewVariables]
         private List<ReagentQuantity> _contents = new List<ReagentQuantity>(2);
+        public List<ReagentQuantity> Contents => _contents;
```
This really needs to be exposed as a read-only interface, like `IReadOnlyList<T>`. You do not want people to try and manually add things to the list, they need to be using the AddReagent function.

Acruid commented on Content.Shared/GameObjects/Components/Chemistry/SolutionComponent.cs:
```diff
@@ -30,6 +30,8 @@ public int MaxVolume
             set => _maxVolume = value; // Note that the contents won't spill out if the capacity is reduced.
         }
 
+        public Solution ContainedSolution => _containedSolution;
```
Making the solution object public means that any code can completely bypass the the SolutionComponent functions when adding or removing a reagent. This is a really bad idea, and breaks general OOP encapsulation rules. You need to design it so that outside code has to ask the SolutionComponent for info about the solution, instead of exposing the internal workings of SolutionComponent for outside code to access.

Moneyl commented:
@Acruid I've implemented those changes. I see what you mean with those, was being lazy. Everything *should* be encapsulated properly to avoid direct access to `SolutionComponent._containedSolution` now. Note that a `List<T>` is passed [here](https://github.com/space-wizards/space-station-14/blob/ad6bf87f0bb3527d1663eb0a5b143fe80ada0c18/Content.Server/GameObjects/Components/Chemistry/ReagentDispenserComponent.cs#L121) instead of a `ReadOnlyCollection<T>`. That's since I was getting a crash in the serializer when passing a `ReadOnlyCollection<T>`. Passing a `List<T>` shouldn't be an issue here anyways since it's a message being sent to the client, and not component data being replicated.

RemieRichards commented on Content.Shared/Chemistry/Solution.cs:
```diff
@@ -17,6 +18,7 @@ public class Solution : IExposeData, IEnumerable<Solution.ReagentQuantity>
         // Most objects on the station hold only 1 or 2 reagents
         [ViewVariables]
         private List<ReagentQuantity> _contents = new List<ReagentQuantity>(2);
+        public ReadOnlyCollection<ReagentQuantity> Contents => _contents.AsReadOnly();
```
Prefer this, it's how we do it in the rest of the codebase.
```suggestion
        public IReadOnlyList<ReagentQuantity> Contents => _contents;
```

RemieRichards commented on Content.Shared/GameObjects/Components/Chemistry/SolutionComponent.cs:
```diff
@@ -52,6 +53,8 @@ public SolutionCaps Capabilities
             set => _capabilities = value;
         }
 
+        public ReadOnlyCollection<Solution.ReagentQuantity> ReagentList => _containedSolution.Contents;
```
Following on from the above,
```suggestion
        public IReadOnlyList<Solution.ReagentQuantity> ReagentList => _containedSolution.Contents;
```

Moneyl commented:
I've made all the changes I intend make with this. As long as there are no more requested changes and everyone's happy with the code it should be ready to merge.

PJB3005 commented on Content.Client/GameObjects/Components/Chemistry/ReagentDispenserBoundUserInterface.cs:
```diff
@@ -0,0 +1,115 @@
+            //Setup window layout/elements
+            _window = new ReagentDispenserWindow()
+            {
+                Title = "Reagent dispenser",
```
You should localize all the strings in here with `ILocalizationManager`.

Moneyl replied:
Done. Did the same for strings in `ReagentDispenserWindow` and any other strings from this PR I could find.

PJB3005 commented on Content.Server/GameObjects/Components/Chemistry/SolutionComponent.cs:
```diff
@@ -76,6 +84,22 @@ protected override void Activate(IEntity user, SolutionComponent component)
+                }
+                else
+                {
+                    message.AddText($"Unknown reagent: {reagent.Quantity}u\n");
```
Localize!

Moneyl replied:
Done.

Moneyl commented:
@Acruid I've implemented those suggestions. Only containers with `SolutionCaps.FitsInDispenser` can be placed in reagent dispensers now. It's a good idea since things like the bucket or mop bucket have ridiculous capacities of 500 and I'd imagine that'd be abused somehow. 

Sidenote: Could maybe be useful to allow them for ghetto chem in the future, and maybe they'd melt or burn depending on their materials and the chems in them but that's beyond this PR.  

Also added colors to the new reagents where it made sense to have one at room temperature.

@PJB3005 Localized any new strings I added that I could find.
